File: mobile-android/app/src/main/java/com/nouvelleeve/pharmacie/ApiClient.kt
package com.nouvelleeve.pharmacie

import android.content.Context
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL

class ApiException(val code: Int, message: String) : Exception(message)

object ApiConfig {
    const val BASE = "https://mapharmaciepk.xo.je/api"
}

class ApiClient(
    private val appContext: Context,
    private val token: String? = null,
    private val sessionId: String? = null
) {
    private val cookieHelper = InfinityFreeCookieHelper(appContext)

    private fun authToken(): String? = token?.trim()?.takeIf { it.isNotBlank() }
    private fun authSessionId(): String? = sessionId?.trim()?.takeIf { it.isNotBlank() }

    private fun withAuthQuery(url: String): String {
        val sb = StringBuilder(url)
        fun appendParam(name: String, value: String) {
            sb.append(if (sb.contains('?')) '&' else '?')
            sb.append(name).append('=').append(encode(value))
        }
        authToken()?.let { appendParam("token", it) }
        authSessionId()?.let { appendParam("sid", it) }
        return sb.toString()
    }

    suspend fun ping(): JSONObject = execute("GET", "${ApiConfig.BASE}/ping.php", null)

    suspend fun login(email: String, password: String): JSONObject {
        return execute(
            "POST",
            "${ApiConfig.BASE}/login.php",
            JSONObject().apply {
                put("email", email)
                put("password", password)
            },
            auth = false
        )
    }

    suspend fun logout(): JSONObject {
        val body = JSONObject().apply {
            authToken()?.let { put("token", it) }
            authSessionId()?.let { put("sid", it) }
        }
        return execute("POST", withAuthQuery("${ApiConfig.BASE}/logout.php"), body)
    }

    suspend fun getMedicaments(query: String = ""): JSONObject {
        val url = if (query.isBlank()) {
            "${ApiConfig.BASE}/medicaments.php"
        } else {
            "${ApiConfig.BASE}/medicaments.php?q=${encode(query)}"
        }
        return execute("GET", withAuthQuery(url), null)
    }

    suspend fun getStock(query: String = ""): JSONObject {
        val url = if (query.isBlank()) {
            "${ApiConfig.BASE}/stock.php"
        } else {
            "${ApiConfig.BASE}/stock.php?q=${encode(query)}"
        }
        return execute("GET", withAuthQuery(url), null)
    }

    suspend fun createVente(
        lignes: List<Triple<Int, Int, Double>>,
        devise: String,
        clientNom: String,
        notes: String = ""
    ): JSONObject {
        val arr = org.json.JSONArray()
        lignes.forEach { (medId, qty, prix) ->
            arr.put(JSONObject().apply {
                put("medicament_id", medId)
                put("quantite", qty)
                if (prix > 0) put("prix_unitaire", prix)
            })
        }
        return execute(
            "POST",
            withAuthQuery("${ApiConfig.BASE}/ventes.php"),
            JSONObject().apply {
                put("lignes", arr)
                put("devise", devise)
                put("client_nom", clientNom)
                if (notes.isNotBlank()) put("notes", notes)
            }
        )
    }

    suspend fun getJournee(date: String? = null): JSONObject {
        val url = buildString {
            append("${ApiConfig.BASE}/journee.php")
            if (!date.isNullOrBlank()) append("?date=${encode(date)}")
        }
        return execute("GET", withAuthQuery(url), null)
    }

    suspend fun getHistoriqueVentes(date: String? = null, limit: Int = 50): JSONObject {
        val params = buildString {
            append("${ApiConfig.BASE}/ventes.php?liste=1&limit=$limit")
            if (!date.isNullOrBlank()) append("&date=${encode(date)}")
        }
        return execute("GET", withAuthQuery(params), null)
    }

    suspend fun getRecu(venteId: Int): JSONObject =
        execute("GET", withAuthQuery("${ApiConfig.BASE}/recu.php?id=$venteId"), null)

    suspend fun rapportJour(date: String): JSONObject =
        execute("GET", withAuthQuery("${ApiConfig.BASE}/rapports.php?type=jour&date=$date"), null)

    suspend fun rapportMois(annee: Int, mois: Int): JSONObject =
        execute("GET", withAuthQuery("${ApiConfig.BASE}/rapports.php?type=mois&annee=$annee&mois=$mois"), null)

    suspend fun getAlertes(type: String = "all"): JSONObject =
        execute("GET", withAuthQuery("${ApiConfig.BASE}/alertes.php?type=${encode(type)}"), null)

    suspend fun getCaisse(date: String? = null): JSONObject {
        val url = buildString {
            append("${ApiConfig.BASE}/caisse.php")
            if (!date.isNullOrBlank()) append("?date=${encode(date)}")
        }
        return execute("GET", withAuthQuery(url), null)
    }

    suspend fun createMouvementCaisse(
        type: String,
        montant: Double,
        devise: String,
        motif: String
    ): JSONObject {
        return execute(
            "POST",
            withAuthQuery("${ApiConfig.BASE}/caisse.php"),
            JSONObject().apply {
                put("type", type)
                put("montant", montant)
                put("devise", devise)
                put("motif", motif)
            }
        )
    }

    private fun encode(value: String): String =
        java.net.URLEncoder.encode(value, Charsets.UTF_8.name())

    private suspend fun execute(
        method: String,
        urlString: String,
        body: JSONObject?,
        auth: Boolean = true
    ): JSONObject = withContext(Dispatchers.IO) {
        withContext(Dispatchers.Main) {
            cookieHelper.ensureCookie()
        }

        val requestBody = when {
            body != null && (authToken() != null || authSessionId() != null) ->
                JSONObject(body.toString()).apply {
                    authToken()?.let { put("token", it) }
                    authSessionId()?.let { put("sid", it) }
                }
            else -> body
        }

        var result = rawRequest(method, urlString, requestBody, auth)

        if (InfinityFreeChallenge.isChallenge(result.body)) {
            if (!InfinityFreeChallenge.solveAndStore(result.body)) {
                withContext(Dispatchers.Main) {
                    cookieHelper.ensureCookie()
                }
            }
            result = rawRequest(method, urlString, requestBody, auth)
        }

        if (InfinityFreeChallenge.isChallenge(result.body)) {
            if (InfinityFreeChallenge.solveAndStore(result.body)) {
                result = rawRequest(method, urlString, requestBody, auth)
            }
        }

        parseResponse(result.code, result.body)
    }

    private data class HttpResult(val code: Int, val body: String)

    private fun rawRequest(method: String, urlString: String, body: JSONObject?, auth: Boolean): HttpResult {
        val conn = (URL(urlString).openConnection() as HttpURLConnection).apply {
            requestMethod = method
            connectTimeout = 30000
            readTimeout = 30000
            setRequestProperty("Content-Type", "application/json; charset=utf-8")
            setRequestProperty("Accept", "application/json")
            setRequestProperty("User-Agent", "Mozilla/5.0 (Linux; Android) AppleWebKit/537.36 Chrome/120.0.0.0 Mobile")
            cookieHelper.cookieHeader()?.let { setRequestProperty("Cookie", it) }
            authToken()?.let { t ->
                setRequestProperty("Authorization", "Bearer $t")
                setRequestProperty("X-Auth-Token", t)
            }
            authSessionId()?.let { sid ->
                setRequestProperty("X-Session-Id", sid)
            }
            doInput = true
            if (body != null && method != "GET") {
                doOutput = true
                OutputStreamWriter(outputStream, Charsets.UTF_8).use { it.write(body.toString()) }
            }
        }

        val responseCode = conn.responseCode
        val stream = if (responseCode in 200..299) conn.inputStream else conn.errorStream
        val text = BufferedReader(InputStreamReader(stream ?: conn.inputStream, Charsets.UTF_8)).use { it.readText() }
        captureSessionCookie(conn)
        conn.disconnect()
        return HttpResult(responseCode, text)
    }

    private fun captureSessionCookie(conn: HttpURLConnection) {
        val setCookies = conn.headerFields?.entries
            ?.filter { it.key.equals("Set-Cookie", ignoreCase = true) }
            ?.flatMap { it.value }
            ?: return

        for (header in setCookies) {
            val match = Regex("PHPSESSID=([^;]+)", RegexOption.IGNORE_CASE).find(header) ?: continue
            val sid = match.groupValues[1]
            if (sid.isNotBlank()) {
                SessionManager(appContext).sessionId = sid
            }
        }
    }

    private fun parseResponse(responseCode: Int, text: String): JSONObject {
        if (InfinityFreeChallenge.isChallenge(text)) {
            throw ApiException(
                responseCode,
                "Connexion bloquée par InfinityFree. Réessayez dans 5 secondes."
            )
        }

        val json = try {
            JSONObject(text)
        } catch (_: Exception) {
            throw ApiException(responseCode, "Réponse serveur invalide ($responseCode)")
        }

        if (!json.optBoolean("success", false)) {
            throw ApiException(responseCode, json.optString("message", "Erreur API"))
        }

        return json.optJSONObject("data") ?: JSONObject()
    }
}

fun JSONArray.toList(): List<JSONObject> {
    val list = mutableListOf<JSONObject>()
    for (i in 0 until length()) {
        list.add(getJSONObject(i))
    }
    return list
}

File: vendeur/js/api-client.js
/**
 * Client API vendeur — même backend que l'APK Android
 */
const Api = {
  base: '/api',

  session() {
    try {
      return JSON.parse(localStorage.getItem('ne_vendeur_session') || '{}');
    } catch {
      return {};
    }
  },

  saveSession(data) {
    localStorage.setItem('ne_vendeur_session', JSON.stringify(data));
  },

  clearSession() {
    localStorage.removeItem('ne_vendeur_session');
  },

  isLoggedIn() {
    const s = this.session();
    return !!(s.token || s.session_id);
  },

  authQuery(url) {
    const s = this.session();
    const u = new URL(url, window.location.origin);
    if (s.token) u.searchParams.set('token', s.token);
    if (s.session_id) u.searchParams.set('sid', s.session_id);
    return u.toString();
  },

  async request(method, path, body = null) {
    const url = path.startsWith('http') ? path : `${this.base}${path}`;
    const finalUrl = method === 'GET' ? this.authQuery(url) : this.authQuery(url);
    const s = this.session();
    const headers = {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    };
    if (s.token) {
      headers.Authorization = `Bearer ${s.token}`;
      headers['X-Auth-Token'] = s.token;
    }
    if (s.session_id) headers['X-Session-Id'] = s.session_id;

    let payload = body;
    if (body && typeof body === 'object') {
      payload = { ...body };
      if (s.token) payload.token = s.token;
      if (s.session_id) payload.sid = s.session_id;
    }

    const res = await fetch(finalUrl, {
      method,
      headers,
      credentials: 'same-origin',
      body: payload ? JSON.stringify(payload) : null,
    });

    const text = await res.text();
    if (text.includes('aes.js')) {
      throw new Error('Connexion bloquée par InfinityFree. Rechargez la page et réessayez.');
    }

    let json;
    try {
      json = JSON.parse(text);
    } catch {
      throw new Error('Réponse serveur invalide.');
    }

    if (!json.success) {
      throw new Error(json.message || 'Erreur API');
    }
    return json.data ?? json;
  },

  login(email, password) {
    return this.request('POST', '/login.php', { email, password });
  },

  logout() {
    return this.request('POST', '/logout.php', {});
  },

  getMedicaments(q = '') {
    const p = q ? `?q=${encodeURIComponent(q)}` : '';
    return this.request('GET', `/medicaments.php${p}`);
  },

  getStock(q = '') {
    const p = q ? `?q=${encodeURIComponent(q)}` : '';
    return this.request('GET', `/stock.php${p}`);
  },

  createVente(data) {
    return this.request('POST', '/ventes.php', data);
  },

  getHistorique(limit = 50) {
    return this.request('GET', `/ventes.php?liste=1&limit=${limit}`);
  },

  getRecu(id) {
    return this.request('GET', `/recu.php?id=${id}`);
  },

  rapportJour(date) {
    return this.request('GET', `/rapports.php?type=jour&date=${date}`);
  },

  getCaisse(date) {
    const p = date ? `?date=${date}` : '';
    return this.request('GET', `/caisse.php${p}`);
  },

  createMouvementCaisse(data) {
    return this.request('POST', '/caisse.php', data);
  },

  getAlertes(type = 'all') {
    return this.request('GET', `/alertes.php?type=${encodeURIComponent(type)}`);
  },
};

const Format = {
  money(amount, devise = 'CDF') {
    const n = Number(amount) || 0;
    if (devise === 'USD') return `$${n.toLocaleString('fr-FR', { minimumFractionDigits: 2, maximumFractionDigits: 2 })}`;
    return `${Math.round(n).toLocaleString('fr-FR')} FC`;
  },
  dateTime(raw) {
    if (!raw) return '—';
    const d = new Date(String(raw).replace(' ', 'T'));
    return isNaN(d) ? raw : d.toLocaleString('fr-FR');
  },
  today() {
    return new Date().toISOString().slice(0, 10);
  },
};
